Run only failed tests

milljm opened the pull request:
Allow the TestHarness to write failed tests to a hidden file.
When running `run_tests` with --failed-tests options, the TestHarness
will only run tests listed in the hidden file.

When ever the user performs a run_tests with out --failed-tests, the
hidden file gets clobbered.

--failed-tests also works along side --re options.

Refs #2285

andrsd commented on python/TestHarness/TestHarness.py:
```diff
@@ -789,6 +804,18 @@ def parseCLArgs(self, argv):
+    # If attempting to test only failed_tests, open the .failed_tests file and create a list object
+    self.test_list = []
+    if self.options.failed_tests:
+      with open(os.path.join(os.getcwd(), '.failed_tests'), 'r') as tmp_failFile:
```
`tmp_failFile` is a mixture of C-style and camel case. It should be C-style -> `tmp_fail_file`.

andrsd commented on python/TestHarness/TestHarness.py:
```diff
@@ -789,6 +804,18 @@ def parseCLArgs(self, argv):
+      try:
+        self.writeFailedTest = open(os.path.join(os.getcwd(), '.failed_tests'), 'w')
+      except IOError:
+        print 'ERROR: Unable to open', os.path.join(os.getcwd(), '.failed_tests'), 'for writing'
```
Use a static variable for the file name...

permcody replied:
Hmm, this is the right place for try/catch but it's debatable that this should be a fatal error. I think it's fine for now because if you can't write to this directory, it's highly unlikely that the tests are going to pass when they start writing output files to each of their directories.

permcody commented on python/TestHarness/TestHarness.py:
```diff
@@ -239,6 +239,10 @@ def findAndRunTests(self, find_only=False):
                 sys.path.pop()
     except KeyboardInterrupt:
+      try:
+        self.writeFailedTest.close()
```
Curious, did this file fail to close during your testing? If not, I'd prefer to remove this try/catch. I'd rather know why something is failing or if it ever fails first. Please remove or document why this is here.

permcody commented on python/TestHarness/TestHarness.py:
```diff
@@ -620,6 +624,10 @@ def handleTestResult(self, specs, output, result, start=0, end=0, add_to_table=T
 
     if not 'skipped' in result:
+      # If we are running normally and a test failes, write the name of the failed test to a file
+      if 'FAILED' in result and not self.options.failed_tests:
```
Note: This will conflict with your other in progress PR.

milljm replied:
Noted.

permcody commented on python/TestHarness/TestHarness.py:
```diff
@@ -687,6 +695,12 @@ def cleanup(self):
 
+    # Close the failed_tests file. Ignore if the file is already closed (ctl-c, keyboard interrupt)
+    try:
+      self.writeFailedTest.close()
```
Ah, now I see. You are worried about closing the file twice? That won't happen in at least one of these cases because we are existing in the other case. That means this one will always work. I suggest that you simply test to see if the file is closed above and then close. try/catch isn't the correct solution here.

permcody commented on python/TestHarness/testers/Tester.py:
```diff
@@ -122,9 +122,14 @@ def processResults(self, moose_dir, retcode, options, output):
+    # Check if we only want to run failed tests
+    if options.failed_tests:
+      if self.specs['test_name'] not in test_list:
+        return (False, '')
```
OK - but this will also conflict with your other PR.

milljm replied:
Noted

permcody reviewed (changes requested):
Overall this looks good. It's the right design, just needs a few fixups.

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit cc49a536e66ef481f6229805656987cb87db6465
+Subproject commit 5129e936af895440f6fc63d117c14d7f02765018
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

permcody commented on python/TestHarness/TestHarness.py:
```diff
@@ -239,6 +242,8 @@ def findAndRunTests(self, find_only=False):
                 sys.path.pop()
     except KeyboardInterrupt:
+      if not self.options.failed_tests:
+        self.writeFailedTest.close()
```
Not quite! You can ask the filehandle directly about it's state, you don't want to use an extra variable to do that because it can easily get out of sync.
```
if (!self.writeFailedTest.closed)
  self.writeFailedTest.close()
```

permcody commented on python/TestHarness/TestHarness.py:
```diff
@@ -789,6 +803,16 @@ def parseCLArgs(self, argv):
+    # If attempting to test only failed_tests, open the .failed_tests file and create a list object
+    # otherwise, open the failed_tests file object for writing (clobber).
+    self.test_list = []
+    failed_tests_file = os.path.join(os.getcwd(), '.failed_tests')
```
Did you test this if you are downstream in a test directory? (e.g. ../../run_tests --re=some_directory). Does it write the failed test to the wrong directory? Instead, use the method we use in `run_tests` to get the location of the script directory, not the cwd.

milljm replied:
This works, because run_tests is changing to the directory when instancing the TestHarness:

```python
# Set the current working directory to the directory where this script is located
os.chdir(os.path.abspath(os.path.dirname(sys.argv[0])))
```

The ClusterLauncher, and PBS also depend on the above assumption. As does the TestHarness os.walk routine:

```python
      else:
        self.options.processingPBS = False
        self.base_dir = os.getcwd()
        for dirpath, dirnames, filenames in os.walk(self.base_dir, followlinks=True):
```

Its probably not the best way, but I was just following suite.

permcody replied:
Oh, OK

permcody reviewed (changes requested):
Almost there...

Commit:
Run only failed tests

Allow the TestHarness to write failed tests to a hidden file.
When running run_tests with --failed-tests options, the TestHarness
will only run tests listed in the hidden file.

When ever the user performs a run_tests with out --failed-tests, the
hidden file gets clobbered.

--failed-tests also works along side --re options.

Refs #2285

permcody reviewed (approved):
I'm approving but I still don't like your new variable name `writeFailedTest`. This seems like an action or possibly a Boolean by the name alone. I better name would be `failed_file_handle`. We can change it at a later time.
